notebook/nbextensions.py

```diff
@@ -299,118 +426,349 @@ def _config_file_name_default(self):
-        ),
-        enable=(EnableNBExtensionApp, "Enable a notebook extension"),
-        disable=(DisableNBExtensionApp, "Disable a notebook extension"),
+        install=(InstallNBExtensionApp,"Install an nbextension"),
+        enable=(EnableNBExtensionApp, "Enable an nbextension"),
+        disable=(DisableNBExtensionApp, "Disable an nbextension"),
+        uninstall=(UninstallNBExtensionApp, "Uninstall an nbextension"),
+        list=(ListNBExtensionsApp, "List nbextensions")
     )
 
     def start(self):
         super(NBExtensionApp, self).start()
 
         # The above should have called a subcommand and raised NoStart; if we
-        # get here, it didn't, so we should print a message.
+        # get here, it didn't, so we should self.log.info a message.
         subcmds = ", ".join(sorted(self.subcommands))
         sys.exit("Please supply at least one subcommand: %s" % subcmds)
 
 main = NBExtensionApp.launch_instance
 
+#------------------------------------------------------------------------------
+# Private API
+#------------------------------------------------------------------------------
+
+
+def _should_copy(src, dest, verbose=1, log=None):
+    """Should a file be copied?"""
+    if not os.path.exists(dest):
+        return True
+    if os.stat(src).st_mtime - os.stat(dest).st_mtime > 1e-6:
+        # we add a fudge factor to work around a bug in python 2.x
+        # that was fixed in python 3.x: http://bugs.python.org/issue12904
+        if verbose >= 2:
+            log("%s is out of date" % dest)
+        return True
+    if verbose >= 2:
+        log("%s is up to date" % dest)
+    return False
+
+
+def _maybe_copy(src, dest, verbose=1, log=None):
+    """Copy a file if it needs updating."""
+    if log is None: log = print
+    if _should_copy(src, dest, verbose=verbose, log=log):
+        if verbose >= 1:
+            log("Copying: %s -> %s" % (src, dest))
+        shutil.copy2(src, dest)
+
+
+def _safe_is_tarfile(path):
+    """Safe version of is_tarfile, return False on IOError."""
+    try:
+        return tarfile.is_tarfile(path)
+    except IOError:
+        return False
+
+
+def _get_nbextension_dir(user=False, sys_prefix=False, prefix=None, nbextensions_dir=None):
+    """Return the nbextension directory specified"""
+    if sum(map(bool, [user, prefix, nbextensions_dir, sys_prefix])) > 1:
+        raise ArgumentConflict("cannot specify more than one of user, sys_prefix, prefix, or nbextensions_dir")
+    if user:
+        nbext = pjoin(jupyter_data_dir(), u'nbextensions')
+    elif sys_prefix:
+        nbext = pjoin(ENV_JUPYTER_PATH[0], u'nbextensions')
+    elif prefix:
+        nbext = pjoin(prefix, 'share', 'jupyter', 'nbextensions')
+    elif nbextensions_dir:
+        nbext = nbextensions_dir
+    else:
+        nbext = pjoin(SYSTEM_JUPYTER_PATH[0], 'nbextensions')
+    return nbext
+
+
+def _get_config_dir(user=False, sys_prefix=False):
+    if sum(map(bool, [user, sys_prefix])) > 1:
```

Comment:
I think this is a long way to write `if user and sys_prefix:`, but it makes sense in the above case with more options.